### PR TITLE
#870: ensureIndexes() throws unhandleable "Index with name: xyz_1" already exists with different options"

### PR DESCRIPTION
**Pull request by UlrichCech**
This is a first implemention for this topic... please take a look here, if this could be a solution and if I handled all situations.

### CODE REVIEW

**Comment by UlrichCech** [comment]
@evanchooly Could to have a look at the current build? I don't get a clearance of this pull-request, because of errors in the CLIRR-report from a former commit.
There was also a findbugs-error in the Group.class which I wanted to address, perhaps you can have a special look at the Group.class of this pull-request... perhaps, this can be reverted or is obsolete.

File: morphia/src/main/java/org/mongodb/morphia/DatastoreImpl.java [hunk 1]
```diff
@@ -1144,12 +1151,58 @@ protected void ensureIndex(final MappedClass mc, final DBCollection dbColl, fina
...
         LOG.debug(format("Creating index for %s with keys:%s and opts:%s", dbColl.getName(), keys, options));
-        dbColl.createIndex(keys, options);
+        createMongoIndex(dbColl, keys, options);
+    }
+
+    private void createMongoIndex(final DBCollection dbColl, final DBObject keys, final DBObject options) {
+        try {
+            dbColl.createIndex(keys, options);
+        } catch (MongoCommandException ex) {
+            if (options.containsField("dropOutdated") && getBoolean(options, "dropOutdated", false)) {
```
Comment (evanchooly) [inline morphia/src/main/java/org/mongodb/morphia/DatastoreImpl.java]: A cast to BasicDBObject is sufficient here and then you'd have access to these methods defined there.

Also, checking contains first is redundant in this case since the default on the get is false.

File: morphia/src/main/java/org/mongodb/morphia/annotations/Index.java [hunk 2]
```diff
@@ -105,6 +105,14 @@
     boolean unique() default false;
 
     /**
+     * Removes an existing index, if the options of the existing index with the same name have changed.
+     *
+     * @deprecated use the {@link IndexOptions} found in {@link #options()}
+     */
+    @Deprecated
+    boolean dropOutdated() default false;
```
Comment (evanchooly) [inline morphia/src/main/java/org/mongodb/morphia/annotations/Index.java]: This shouldn't be here.  It doesn't make sense to add something that's deprecrated.  New features should go on the new code only.

File: morphia/src/main/java/org/mongodb/morphia/annotations/Indexed.java [hunk 3]
```diff
@@ -76,6 +76,14 @@
     boolean unique() default false;
 
     /**
+     * Removes an existing index, if the options of the existing index with the same name have changed.
+     *
+     * @deprecated use the {@link IndexOptions} found in {@link #options()}
+     */
+    @Deprecated
+    boolean dropOutdated() default false;
```
Comment (evanchooly) [inline morphia/src/main/java/org/mongodb/morphia/annotations/Indexed.java]: Again, this value should only be `IndexOptions`

File: morphia/src/main/java/org/mongodb/morphia/aggregation/Group.java [hunk 4]
```diff
@@ -233,7 +234,12 @@ public String getSourceField() {
      * @return the projections for the group
      */
     public Projection[] getProjections() {
-        return projections;
+        if (projections != null) {
```
Comment (evanchooly) [inline morphia/src/main/java/org/mongodb/morphia/aggregation/Group.java]: This doesn't belong in this PR and has been fixed elsewhere.

**Comment by UlrichCech** [comment]
@evanchooly  I don't get the PR green on Travis... it says, that there are 3 test-failures, but I don't know where... if I run the test-suite locally, everything is green.

**Comment by evanchooly** [comment]
master is green too.  try rebasing and force pushing an update to the PR.  Or creating a new one.

**Comment by UlrichCech** [comment]
I try with a new PR
